ViewController.swift

```diff
@@ -7,13 +7,100 @@
 
 import UIKit
 
-class ViewController: UIViewController {
-
+class ViewController: UICollectionViewController, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
+    var people = [Person]()
+    
     override func viewDidLoad() {
         super.viewDidLoad()
-        // Do any additional setup after loading the view.
+        navigationItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(addNewPerson))
     }
 
-
+    override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return people.count
+    }
+    
+    override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "Person", for: indexPath) as? PersonCell else {
+            fatalError("Enable to deque PersonCell")
+        }
+        let person = people[indexPath.item]
+        cell.name.text = person.name
+        let path = getDocumentsDirectory().appendingPathComponent(person.image)
+        cell.imageView.image = UIImage(contentsOfFile: path.path)
+        cell.imageView.layer.borderColor = UIColor(white: 0, alpha: 0.3).cgColor
+        cell.imageView.layer.borderWidth = 2
+        cell.imageView.layer.cornerRadius = 3
+        cell.layer.cornerRadius = 7
+        
+        return cell
+    }
+    
+    @objc private func addNewPerson() {
+        let picker = UIImagePickerController()
+        picker.allowsEditing = true
+        picker.delegate = self
+        present(picker, animated: true)
+    }
+    
+    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
+        guard let image = info[.editedImage] as? UIImage else { return }
+        let imageName = UUID().uuidString
+        let imagepath = getDocumentsDirectory().appendingPathComponent(imageName)
+        
+        if let jpegData = image.jpegData(compressionQuality: 0.8) {
+            try? jpegData.write(to: imagepath)
+        }
+        let person = Person(name: "Unknown", image: imageName)
+        people.append(person)
+        collectionView.reloadData()
+        dismiss(animated: true)
+    }
+    
+    private func getDocumentsDirectory() -> URL {
+        let paths = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
+        return paths[0]
+    }
+    
+    override func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        let person = people[indexPath.item]
+        
+        let alertController = UIAlertController(title: nil, message: "Please Select an Option", preferredStyle: .actionSheet)
+        
+        alertController.addAction(UIAlertAction(title: "Rename", style: .default, handler: { (_) in
```

Comment:
If you want to avoid this chain of closures (which is hard to read) you can just move the handler code to another function. This function could be a factory method, which would make the rename alert controller for you. It would be something like:

```
func makeRenameAlertViewController() -> UIAlertViewController {
    // [your code to make the alert]
}
```

Then in the handler, you'd just invoke it and present it, like so:

```suggestion
        alertController.addAction(UIAlertAction(title: "Rename", style: .default, handler: { [weak self] _ in 
            guard let self = self else { return }
            let renameAlertViewController = self.makeRenameAlertViewController()
            self.present(renameAlertViewController, animated: true)
        }
```

*Please note the parenthesis are not needed for parameters in closures unless you declare their types.

Reply:
Awesome tip!!
I just had to pass the indexPath to the function. 
When I first implemented this I spent some time trying to make it more readable. 
Very good!! Fixed!